src/Bicep.Core/Emit/SourceMapHelper.cs

```diff
@@ -0,0 +1,235 @@
+            TextSpan bicepLocation = bicepSyntax.Span;
+
+            // account for leading nodes (decorators)
+            if (bicepSyntax is StatementSyntax syntax)
+            {
+                var lastLeadingNode = syntax.LeadingNodes.LastOrDefault();
+                if (lastLeadingNode is not null)
+                {
+                    bicepLocation = new TextSpan(
+                        lastLeadingNode.Span.Position + lastLeadingNode.Span.Length,
+                        syntax.Span.Length - lastLeadingNode.Span.Length);
+                }
+            }
+
+            // increment start position if starting on a comma (happens when outputting successive items in objects and arrays)
+            if (jsonWriter.CommaPositions.Contains(jsonStartPos))
+            {
+                jsonStartPos++;
+            }
+
+            var jsonEndPos = jsonWriter.CurrentPos - 1;
+            var jsonPos = new TextSpan(jsonStartPos, jsonEndPos - jsonStartPos);
+
+            rawSourceMap.AddMapping(
+                bicepFilePath,
+                bicepLocation,
+                jsonPos);
+        }
+
+        public static void AddNestedRawSourceMap(
+            this RawSourceMap parentSourceMap,
+            RawSourceMap nestedSourceMap,
+            int offset)
+        {
+            foreach (var fileEntry in nestedSourceMap.Entries)
+            {
+                foreach (var sourceMapEntry in fileEntry.SourceMap)
+                {
+                    foreach (var position in sourceMapEntry.TargetPositions)
+                    {
+                        var positionWithOffset = new TextSpan(
+                            position.Position + offset,
+                            position.Length);
+
+                        parentSourceMap.AddMapping(
+                            fileEntry.FilePath,
+                            sourceMapEntry.SourcePosition,
+                            positionWithOffset);
+                    }
+                }
+            }
+        }
+
+        public static SourceMap ProcessRawSourceMap(
+            RawSourceMap rawSourceMap,
+            JToken rawTemplate,
+            BicepFile sourceFile)
+        {
+            var formattedTemplateLines = rawTemplate
+                .ToString(Formatting.Indented)
+                .Split(Environment.NewLine, StringSplitOptions.None);
+
+            // get line starts of unformatted JSON by stripping formatting from each line of formatted JSON
+            var unformattedLineStarts = formattedTemplateLines
+                .Aggregate(
+                    new List<int>() { 0 }, // first line starts at position 0
+                    (lineStarts, line) =>
+                    {
+                        var unformattedLine = JsonWhitespaceStrippingRegex.Replace(line, "$1");
+                        lineStarts.Add(lineStarts.Last() + unformattedLine.Length);
+                        return lineStarts;
```

Comment:
Why is this logic to strip whitespace at the start of the lines necessary? Shouldn't the conversion between offset <=> coordinates be agnostic to whitespace?

Reply:
During the templating writing, the offsets for the JSON being recorded are for the raw, unformatted JSON. But the final source map's target line numbers are the line numbers of the formatted JSON. The act of formatting the JSON introduces whitespace that invalidates all the offsets that have already been recorded for the raw, unformatted JSON.

So, by taking the lines of the formatted JSON, stripping all whitespace not in strings (so more than just starting whitespace), we can effectively get the line starts of the raw JSON (e.g. 
`string.Join(strippedJsonLines) == rawUnformattedJson`). Then we use these to get the final line numbers of the JSON from the "raw" offsets we have in the source map.

If you can think of a more elegant solution to this, I would love to implement it. 👍